File: 61.rotate-list.py
#
# @lc app=leetcode id=61 lang=python3
#
# [61] Rotate List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:

        if not head or not head.next:
            return head
        
        list_length = 1
        tail = head

        while tail.next:
            list_length += 1
            tail = tail.next
        
        tail.next = head

     
        for _ in range(list_length - k % list_length):
            tail = tail.next
        
        head = tail.next
        tail.next = None

        return head


        
# @lc code=end

File: 61.rotate-list.java
/*
 * @lc app=leetcode id=61 lang=java
 *
 * [61] Rotate List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode tail = head;
        int listLength = 1;

        while (tail.next != null) {
            listLength++;
            tail = tail.next;
        }

        tail.next = head;

        for (int i = 0; i < listLength - k % listLength; i++) {
            tail = tail.next;
        }

        head = tail.next;
        tail.next = null;

        return head;
    }
}
// @lc code=end

